x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedUpdate.java

```diff
@@ -140,6 +144,12 @@ public DatafeedUpdate(StreamInput in) throws IOException {
         this.scrollSize = in.readOptionalVInt();
         this.chunkingConfig = in.readOptionalWriteable(ChunkingConfig::new);
         delayedDataCheckConfig = in.readOptionalWriteable(DelayedDataCheckConfig::new);
+        // TODO: change version in backport
+        if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
+            maxEmptySearches = in.readOptionalInt();
```

Comment:
Why do we use `readOptionalInt` here and `readOptionalVInt` in `DatafeedConfig`?

Reply:
A `VInt` cannot be negative.  The datafeed config object can store "never stop this datafeed due to lack of data" as `null`, so can map an input of `-1` to a stored `null`.  So what is stored in the config is either `null` or a positive integer.  But the update needs to distinguish "I don't want to update this setting" from "I want to update this setting to 'never'".  I'm using `null` for "I don't want to update this setting" and `-1` for "I want to update this setting to 'never'". But this means that the `-1` needs to be stored within the update and potentially transported to a different node, so it cannot be a `VInt`.

I suppose I could have used `0` to encode "I want to update this setting to 'never'" as `0` is not a valid setting.  But elsewhere in Elasticsearch `-1` is a more common way of saying "never".  And I could have made the internal representation `0` even though the external configuration is `-1`, but the detrimental effect on tracing user-input values through the code seems to outweigh the slight benefit in number of bytes transmitted over the network.

Reply:
Understand, thanks for detailed explanation.